### ISSUE TITLE
Loading gold standard subjects as an evolving fraction of total subjects in queue

### ISSUE THREAD

**Issue by trouille** [zooniverse/Panoptes#1895]
Deadline: ~Aug 24th

This is a specific need now for the Gravity Spy Project Builder project, but other projects like moving Space Warps onto the Project Builder will want it as well. 

Talking with @chrislintott, the idea is to first hard code this in for Gravity Spy (https://www.zooniverse.org/lab/1104). Eventually we'll create a UI in the Project Builder for research teams to turn this on, but not for now. 

For Gravity Spy, the request is:
- mean frequency of 2/5 gold standard images for the first ~20 subjects
- drop that mean frequency from ~40% to ~30% for the second 20 subjects, ~20% for the third 20 subjects, and the minimum rate of 10 per cent after that.
- this follows the equation: fraction = 2/(5 × 2(int(N_c /20)+1)/2 )

We are not tied specifically to that equation. If it adds complication to do it based on the equation, it would be 100% ok to instead find another way to drop the frequency over time. Whatever makes most sense in the code for you. 

The equation is from the Space Warps logic from their article: https://mnras.oxfordjournals.org/content/455/2/1171.short?cited-by=yes&legid=mnras;455/2/1171

Pinging @chrislintott as heads up.

**Comment by trouille** [comment]
A difference between Gravity Spy and Space Warps is that while Space Warps has only one workflow, Gravity Spy has several. As a volunteer works within Gravity Spy, s/he is promoted to the next workflow level of difficulty, from beginner levels (B1-B4) to apprentice to master (words we use only internally with the research team). 

We will want the counter on the number of classifications a volunteer has done to reset when they move from one workflow level to the next in Gravity Spy. I.e., we'll want them to see 2/5 gold standard images when they start Beginner Level 1, and again when they start Beginner Level 2 workflow, and again when they start the Beginner Level 3 workflow, etc. Will that be possible @marten ?

**Comment by trouille** [comment]
Also note:

The research team has created gold standard subject sets for each class of glitch. 

Beginner level 1 workflow would have the blip gold standard subject set and the whistle gold standard subject set. 

Beginner level 2 workflow would have the blip gold standard subject set, the whistle gold standard subject set, the koi fish gold standard subject, and the power line gold standard subject set. 

And so on. 

Can you work with a set up like this?

**Comment by marten** [comment]
Will there be lots of gold standard images, or will we also need to do something special so that gold standard subjects don't retire?

**Comment by trouille** [comment]
@marten Good question. Best to not have the gold standard subjects retire. Some glitch classes will likely have a couple hundred gold standard subjects, but some will have fewer than 100.

**Comment by chrislintott** [comment]
If that adds complication, couldn't we just create lots of duplicates of the gold standard set?

**Comment by marten** [comment]
Not really, we'll just only use Nero for retirement and that'll be configured to skip gold masters and do a normal numeric limit for the rest.

**Comment by camallen** [comment]
Just a follow up - @marten is working on a backend solution for this, we'll have more information next week.

**Comment by marten** [comment]
As for the rest: @camallen and I came up with the following. Panoptes is already set up to have subject selection strategies beyond the built-in one. The current other one is Cellect. For this, we're building another one that will mix subject sets at given ratios, and we'll add that formula so that the ratios depend on the number of seen subjects. 

We think this would be the nicest solution in the end (working towards making selection pluggable for projects), but because we'll be building a new service, we've time-boxed it to the end of the week, at which point we'll evaluate if it's not just taking too much time.

Possible fallback plans are:

* Let the front-end randomly inject gold-masters. That doesn't really scale well if there's loads of gold masters.
* Use grouped subject selection and build stuff around that
* Launch with just a static ratio, which we can maintain using Nero.

@camallen Did I forget any fallbacks?
@trouille Is that 24th a hard deadline or are there likely to be other kinds of delays too?

**Comment by camallen** [comment]
the only caveat fallback i have is that if this new service falls over / doesn't respond then the default database selector will have to rely on a baseline distribution of GS / normal subjects manually set in the workflow.

**Comment by trouille** [comment]
We're looking on track for other pieces. The hardest is whether Zach + Sarah will have the promotion pieces in place for testing by the 24th as well. Likely.

**Comment by trouille** [comment]
@camallen Could you update on where things stand/look with this? We're looking on track for the 24th to have the other pieces in place for testing. The goal is a Sept 5th beta test start.

**Comment by camallen** [comment]
@trouille i think we're on track for the 24th but i'll defer to @marten as he's building the new selection service.

**Comment by marten** [comment]
@trouille @camallen Okay, so I've got the service to handle that ratio developed (https://github.com/zooniverse/cellect_ex). There's still a bunch of TODOs for this though:

* Deploying that service
* Wiring up Panoptes to use it
* Testing the complete integrated system
* Setting up a Nero script so that gold standard doesn't get retired.

Problem is that I'm out to Salzburg for meetings next week monday-wednesday. Not sure how much I'll get done while there.

**Comment by camallen** [comment]
understood, I can help deploy and wire up the panoptes client to use it.

**Comment by marten** [comment]
(But we should be able to make it in time for the beta test on Sept 5, which I missed initially. I thought that was what was planned for 24th.)

**Comment by trouille** [comment]
The 24th is our internal deadline. We've learned from experience that we really need 10 days buffer and testing time to shake out all the bugs for projects like this, before our planned Beta test deadline. 

Is the 29th feasible for you? That would provide 5 days buffer before beta test for this particular piece in the mix.

**Comment by trouille** [comment]
@mcbouslog Do you have the info you need to know how to continue making progress on the gold standard feedback, if this isn't 100% in place till the 29th?

**Comment by mcbouslog** [comment]
tl;dr - I'm good.

I should have something working this afternoon, then will request review to make sure I'm on the right page. If my general understanding is correct, I'm good regardless of Marten's work, as what I'm building depends purely on the individual subject and whether it includes the gold standard info in it's metadata. I don't think the ratio of when given, how and not retiring the gold standard subjects affects my little piece directly, but please let me know if I'm mistaken, or when I share my work later today it should confirm.

**Comment by marten** [comment]
There shouldnt be any dependencies between our work. 

- Marten
